cuqi/geometry/_geometry.py

```diff
@@ -702,32 +702,60 @@ def _compute_eigpairs(self, xnod, C_nu, sigma, M, N_GL):
 
 class StepExpansion(Continuous1D):
     '''
-    class representation of the step random field with 3 intermidiate steps
+    Class representation of step functions (piecewise constant functions) with `n_steps` 
+    equidistant steps on the interval [0, L].
+    The function `par2fun` maps the parameters `p` (which are the step magnitudes) to the 
+    corresponding step function evaluated on the spacial grid (`grid`) of nodes x0=0, x1, ...xn=L.
+    
+    For example, if `n_steps` is 3 and `grid` is a uniform grid with nodes x0=0, x1=0.1L, ..., xn=L, 
+    then the resulting function evaluated on the grid will be p[0], if x<=L/3, p[1], if L/3<x<=2L/3, 
+    p[2], if 2L/3<x<=L.
+    
+    Parameters:
+    -----------
+    grid: ndarray
+        Grid points for the step expansion to be evaluated at. There can be more grid points 
+        than number of steps, which can be useful, for example, when using the StepExpansion 
+        geometry as a domain geometry for a cuqipy :class:`Model` that expects the input to be
+        interpolated on a (possibly fine) grid (`grid`).
+
+    n_steps: int
+        Number of equidistant steps.
+
+    kwargs: keyword arguments
+        keyword arguments are passed to the initializer of :class:`~cuqi.geometry.Continuous1D`
     '''
-    def __init__(self, grid, axis_labels=['x'], **kwargs):
+    def __init__(self, grid, n_steps=3, **kwargs):
 
-        super().__init__(grid, axis_labels,**kwargs)
+        super().__init__(grid, **kwargs)
+        if n_steps > np.size(self.grid):
+            raise ValueError("n_steps must be smaller than the number of grid points")
 
+        self._n_steps = n_steps
         L = self.grid[-1]
-        self._idx1 = np.where( (self.grid>=0*L)&(self.grid<=0.333*L) )
-        self._idx2 = np.where( (self.grid>0.333*L)&(self.grid<=0.666*L) )
-        self._idx3 = np.where( (self.grid>0.666*L)&(self.grid<=L) )
+
+        self._indices = []
+        for i in range(self._n_steps):
+            start = i*L/self._n_steps
+            end = (i+1)*L/self._n_steps
+            if i ==0:
+                self._indices.append( np.where((self.grid>=start)&(self.grid<=end))[0] )
+            else:
+                self._indices.append( np.where((self.grid>start)&(self.grid<=end))[0] )
 
     def par2fun(self, p):
         real = np.zeros_like(self.grid)  
- 
-        real[self._idx1[0]] = p[0]
-        real[self._idx2[0]] = p[1]
-        real[self._idx3[0]] = p[2]
+        for i in range(self._n_steps):
+            real[self._indices[i]] = p[i]
  
         return real
 
-    def fun2par(self,f):       
-        val1 = f[self._idx1[0][0]]
-        val2 = f[self._idx2[0][0]]
-        val3 = f[self._idx3[0][0]]
-        return np.array([val1,val2,val3])
+    def fun2par(self,f):
+        val = np.zeros(self._n_steps)
+        for i in range(self._n_steps):
+            val[i] = f[self._indices[i][0]]
```

Comment:
I wonder if taking the first element of the list is really the correct/best way of converting from fun to par. Seems to assume we have a function vector that is already constant on each step, which may not be the case. Should one perhaps instead average over the values within the step? This might be seen as some kind of projection on to the step expansion basis functions.

Reply:
This is a very good point! I kept it like this at first to keep the logic unchanged, but now I updated this to support mean, max, min, projection types.